src/server/transaction.cc

```diff
@@ -235,16 +139,127 @@ OpStatus Transaction::InitByArgs(DbIndex index, CmdArgList args) {
 
     ++unique_shard_cnt_;
     unique_shard_id_ = i;
+
     for (size_t j = 0; j < si.args.size(); ++j) {
       *next_arg = si.args[j];
-      if (needs_reverse_mapping) {
+      if (rev_mapping) {
         *rev_indx_it++ = si.original_index[j];
       }
       ++next_arg;
     }
   }
 
   CHECK(next_arg == args_.end());
+}
+
+void Transaction::InitMultiData(KeyIndex key_index) {
+  DCHECK(multi_);
+  auto args = cmd_with_full_args_;
+
+  // TODO: to determine correctly locking mode for transactions, scripts
+  // and regular commands.
+  IntentLock::Mode mode = Mode();
+  multi_->keys.clear();
+  tmp_space.uniq_keys.clear();
+
+  auto lock_func = [this, mode](auto key) {
+    if (tmp_space.uniq_keys.insert(key).second) {
+      if (multi_->is_expanding) {
+        multi_->keys.push_back(key);
+      } else {
+        multi_->lock_counts[key][mode]++;
+      }
+    }
+  };
+
+  // With EVAL, we call this function for EVAL itself as well as for each command
+  // for eval. currently, we lock everything only during the eval call.
+  if (multi_->is_expanding || !multi_->locks_recorded) {
+    for (size_t i = key_index.start; i < key_index.end; i += key_index.step)
+      lock_func(ArgS(args, i));
+    if (key_index.bonus > 0)
+      lock_func(ArgS(args, key_index.bonus));
+  }
+
+  multi_->locks_recorded = true;
+}
+
+void Transaction::StoreKeysInArgs(KeyIndex key_index, bool rev_mapping) {
+  auto args = cmd_with_full_args_;
+
+  // even for a single key we may have multiple arguments per key (MSET).
+  for (unsigned j = key_index.start; j < key_index.start + key_index.step; ++j) {
+    args_.push_back(ArgS(args, j));
+  }
+
+  if (rev_mapping) {
+    reverse_index_.resize(args_.size());
+    for (unsigned j = 0; j < reverse_index_.size(); ++j) {
+      reverse_index_[j] = j + key_index.start - 1;
+    }
+  }
+}
+
+/**
+ *
+ * There are 4 options that we consider here:
+ * a. T spans a single shard and its not multi.
```

Comment:
use consistent notion for transaction: for example, `Tx` instead of T and Trans

Reply:
I didn't touch you comments at all yet

Reply:
oh, I thought their are yours. in that case the comments are perfect 🙃